lib/srv/discovery/discovery.go

```diff
@@ -782,13 +794,79 @@ func (s *Server) handleEC2Instances(instances *server.EC2Instances) error {
 	// instances.Rotation is true whenever the instances received need
 	// to be rotated, we don't want to filter out existing OpenSSH nodes as
 	// they all need to have the command run on them
-	if !instances.Rotation {
+	//
+	// Integration/EICE Nodes don't have heartbeat.
+	// Those Nodes must not be filtered, so that we can extend their expiration and sync labels.
+	if !instances.Rotation && instances.Integration == "" {
 		s.filterExistingEC2Nodes(instances)
 	}
 	if len(instances.Instances) == 0 {
 		return trace.NotFound("all fetched nodes already enrolled")
 	}
 
+	switch {
+	case instances.Integration != "":
+		s.handleEC2UsingEICE(instances)
+	default:
+		if err := s.handleEC2RemoteInstallation(instances); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	if err := s.emitUsageEvents(instances.MakeEvents()); err != nil {
+		s.Log.WithError(err).Debug("Error emitting usage event.")
+	}
+
+	return nil
+}
+
+// handleEC2UsingEICE creates a Teleport Node (EICE subkind) from the instances list.
+func (s *Server) handleEC2UsingEICE(instances *server.EC2Instances) {
+	awsInfo := &types.AWSInfo{
+		AccountID:   instances.AccountID,
+		Region:      instances.Region,
+		Integration: instances.Integration,
+	}
+
+	existingEICENodes := s.existingEICENodes()
+
+	// Add EC2 Instances using EICE method
+	for _, ec2Instance := range instances.Instances {
```

Comment:
I don't think this logic will scale well. This will basically be upserting each discovered EIC instance full node resource on every discovery iteration. Imagine an account with 10-20K or more instances and this will probably decimate the backend.

So similar to my comment below I still think we should make this work more like a heartbeat where we only update the full node if something changed and only bump TTL (keep-alive) otherwise.

Would be helpful to get scale team input on this as well. @fspmarshall @rosstimothy What would you guys recommend here to make this more scalable?

Reply:
> So similar to my comment below I still think we should make this work more like a heartbeat where we only update the full node if something changed and only bump TTL (keep-alive) otherwise.

The `Backend.KeepAlive` method is kind of deceptive.  For the majority of our backends `KeepAlive` isn't any cheaper than a `Put`.  IMO trying to optimize something by switching from a write to a keepalive isn't useful.

> What would you guys recommend here to make this more scalable?

Skipping unnecessary writes (e.g. for a resource that is unchanged) is an excellent optimization.  The _most_ important thing to keep in mind is that the worst thing for scalability is _spikes_ in load.  The same number of writes spread out smoothly over a few minutes rather than being fired off one after the other is _much_ better.

We've tackled similar questions before (e.g. the label reconciler), and we're fairly satisfied with the solution used there.  Basically, instead of writing everything at once, spread out the writes into lots of small batches over time, and use a dynamically calculated batch size to ensure that batches are small, but that all writes still complete within a "worst case" time.  You can see the exact logic used in `discovery/reconciler.go`, but the basic concept is this:

Pick a fairly small "default batch size" (ideally less than `10`, label reconciler uses `5` which I think is a good default).  Pick a batch write interval (ideally about `1s`). Pick a "worst case time" for full sync (longer is better, label reconciler uses `15m`).  Then calculate your batch size as the maximum of either your min batch size or the number of items to sync divided by the number of write intervals within your worst-case time.  Ex:

```golang
const worstCaseSyncTime = time.Minute * 15
const batchWriteInterval = time.Second
const minBatchSize = 5
func batchSize(itemsToSync int) int {
	maxBatches := int(worstCaseSyncTime / batchWriteInterval)
	dynamicBatchSize := (itemsToSync / maxBatches) + 1
	if dynamicBatchSize < minBatchSize {
		return minBatchSize
	}
	return dynamicBatchSize
}
```

The above results in 5 writes per second up to a few thousands "items to sync", then the batch size starts scaling up in order to ensure that the full sync continues to occur within the target worst-case sync time (batch size of 10 at ~8k resources, batch size of 20 at ~17k resources, batch size of 40 at ~35k resources, etc).

It's not a perfect solution (something a little more mathematically complex would probably provide a bit nicer of a scaling curve), but it's decent enough I think.

Reply:
I added that in https://github.com/gravitational/teleport/pull/35504/commits/748e8b42e44f9f70c0694fc27ce4baaae61d4422

What do you think about the implementation?

Reply:
Looks reasonable to me!  I like the idea of creating a separate helper for this kind of thing that we can improve over time.